### DISCUSSION TITLE
[SDK] <Title>zksync2 cannot be installed because of a pysha3 error

### DISCUSSION THREAD

**Discussion by baalisgood** [zkSync-Community-Hub/zksync-developers#122] [SDKs] [answered]
### Environment

Mainnet

### Acknowledgement

- [ ] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

There was an error when I installed the zksync2 dependency with pip, requiring me to install pysha3, but pysha3 has inherited to python6 and I cannot install pysha3，pysha3 code base prompts“The pysha3 package has reached its end of life and is no longer supported.The SHA-3 code was merged into Python 3.6 many years ago.   Please use SHA-3 functions from hash lib module directly.”
So now I'm stuck，I need help.

### Expected Behavior

pip install zksync2

### Code Example

(venv) PS F:\web3> pip install zksync2
Looking in indexes: https://mirrors.aliyun.com/pypi/simple/
Collecting zksync2
  Downloading https://mirrors.aliyun.com/pypi/packages/23/8f/a4da74f4823b7a454a7c99cdd4cad94d7c4fcf1777842f6a095281ef2698/zksync2-0.6.0-py3-none-an
y.whl (43 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.9/43.9 kB 2.2 MB/s eta 0:00:00
Collecting eip712-structs==1.1.0 (from zksync2)
  Using cached eip712_structs-1.1.0-py3-none-any.whl
Collecting web3==6.0.0 (from zksync2)
  Downloading https://mirrors.aliyun.com/pypi/packages/ca/bf/d63c5aa4077e0a531f594391b4b58f49036f50009bc90a30b680b8577e14/web3-6.0.0-py3-none-any.w
hl (568 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 569.0/569.0 kB 5.1 MB/s eta 0:00:00
Requirement already satisfied: eth-utils>=1.4.0 in f:\web3\venv\lib\site-packages (from eip712-structs==1.1.0->zksync2) (2.2.0)
Collecting pysha3>=1.0.2 (from eip712-structs==1.1.0->zksync2)
  Using cached https://mirrors.aliyun.com/pypi/packages/73/bf/978d424ac6c9076d73b8fdc8ab8ad46f98af0c34669d736b1d83c758afee/pysha3-1.0.2.tar.gz (829
 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: aiohttp>=3.7.4.post0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (3.8.5)
Requirement already satisfied: eth-abi>=4.0.0-b.2 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (4.2.0)
Requirement already satisfied: parsimonious==0.9.0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (0.9.0)
Requirement already satisfied: eth-account>=0.8.0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (0.9.0)
Requirement already satisfied: eth-hash[pycryptodome]>=0.5.1 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (0.5.2)
Requirement already satisfied: eth-typing>=3.0.0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (3.4.0)
Requirement already satisfied: hexbytes>=0.1.0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (0.3.1)
Requirement already satisfied: jsonschema>=4.0.0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (4.19.0)
Requirement already satisfied: lru-dict>=1.1.6 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (1.2.0)
Requirement already satisfied: protobuf>=4.21.6 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (4.24.2)
Requirement already satisfied: requests>=2.16.0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (2.31.0)
Requirement already satisfied: websockets>=10.0.0 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (11.0.3)
Requirement already satisfied: pywin32>=223 in f:\web3\venv\lib\site-packages (from web3==6.0.0->zksync2) (306)
Requirement already satisfied: regex>=2022.3.15 in f:\web3\venv\lib\site-packages (from parsimonious==0.9.0->web3==6.0.0->zksync2) (2023.8.8)      
Requirement already satisfied: attrs>=17.3.0 in f:\web3\venv\lib\site-packages (from aiohttp>=3.7.4.post0->web3==6.0.0->zksync2) (23.1.0)
Requirement already satisfied: charset-normalizer<4.0,>=2.0 in f:\web3\venv\lib\site-packages (from aiohttp>=3.7.4.post0->web3==6.0.0->zksync2) (3.
2.0)
Requirement already satisfied: multidict<7.0,>=4.5 in f:\web3\venv\lib\site-packages (from aiohttp>=3.7.4.post0->web3==6.0.0->zksync2) (6.0.4)
Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in f:\web3\venv\lib\site-packages (from aiohttp>=3.7.4.post0->web3==6.0.0->zksync2) (4.0
.3)
Requirement already satisfied: yarl<2.0,>=1.0 in f:\web3\venv\lib\site-packages (from aiohttp>=3.7.4.post0->web3==6.0.0->zksync2) (1.9.2)
Requirement already satisfied: frozenlist>=1.1.1 in f:\web3\venv\lib\site-packages (from aiohttp>=3.7.4.post0->web3==6.0.0->zksync2) (1.4.0)       
Requirement already satisfied: aiosignal>=1.1.2 in f:\web3\venv\lib\site-packages (from aiohttp>=3.7.4.post0->web3==6.0.0->zksync2) (1.3.1)        
Requirement already satisfied: bitarray>=2.4.0 in f:\web3\venv\lib\site-packages (from eth-account>=0.8.0->web3==6.0.0->zksync2) (2.8.1)
Requirement already satisfied: eth-keyfile>=0.6.0 in f:\web3\venv\lib\site-packages (from eth-account>=0.8.0->web3==6.0.0->zksync2) (0.6.1)
Requirement already satisfied: eth-keys>=0.4.0 in f:\web3\venv\lib\site-packages (from eth-account>=0.8.0->web3==6.0.0->zksync2) (0.4.0)
Requirement already satisfied: eth-rlp>=0.3.0 in f:\web3\venv\lib\site-packages (from eth-account>=0.8.0->web3==6.0.0->zksync2) (0.3.0)
Requirement already satisfied: rlp>=1.0.0 in f:\web3\venv\lib\site-packages (from eth-account>=0.8.0->web3==6.0.0->zksync2) (3.0.0)
Requirement already satisfied: pycryptodome<4,>=3.6.6 in f:\web3\venv\lib\site-packages (from eth-hash[pycryptodome]>=0.5.1->web3==6.0.0->zksync2) 
(3.18.0)
Requirement already satisfied: cytoolz>=0.10.1 in f:\web3\venv\lib\site-packages (from eth-utils>=1.4.0->eip712-structs==1.1.0->zksync2) (0.12.2)
Requirement already satisfied: jsonschema-specifications>=2023.03.6 in f:\web3\venv\lib\site-packages (from jsonschema>=4.0.0->web3==6.0.0->zksync2
) (2023.7.1)
Requirement already satisfied: referencing>=0.28.4 in f:\web3\venv\lib\site-packages (from jsonschema>=4.0.0->web3==6.0.0->zksync2) (0.30.2)
Requirement already satisfied: rpds-py>=0.7.1 in f:\web3\venv\lib\site-packages (from jsonschema>=4.0.0->web3==6.0.0->zksync2) (0.10.0)
Requirement already satisfied: idna<4,>=2.5 in f:\web3\venv\lib\site-packages (from requests>=2.16.0->web3==6.0.0->zksync2) (3.4)
Requirement already satisfied: urllib3<3,>=1.21.1 in f:\web3\venv\lib\site-packages (from requests>=2.16.0->web3==6.0.0->zksync2) (2.0.4)
Requirement already satisfied: certifi>=2017.4.17 in f:\web3\venv\lib\site-packages (from requests>=2.16.0->web3==6.0.0->zksync2) (2023.7.22)      
Requirement already satisfied: toolz>=0.8.0 in f:\web3\venv\lib\site-packages (from cytoolz>=0.10.1->eth-utils>=1.4.0->eip712-structs==1.1.0->zksyn
c2) (0.12.0)
Building wheels for collected packages: pysha3
  Building wheel for pysha3 (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [16 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-311
      copying sha3.py -> build\lib.win-amd64-cpython-311
      running build_ext
      building '_pysha3' extension
      creating build\temp.win-amd64-cpython-311
      creating build\temp.win-amd64-cpython-311\Release
      creating build\temp.win-amd64-cpython-311\Release\Modules
      creating build\temp.win-amd64-cpython-311\Release\Modules\_sha3
      "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DND
EBUG /MD -DPY_WITH_KECCAK=1 -IF:\web3\venv\include "-IC:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1520.0_x64__qbz5n2kfra
8p0\include" "-IC:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1520.0_x64__qbz5n2kfra8p0\Include" "-IC:\Program Files (x86)
\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Au
xiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.20348.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.20
348.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.20348.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.20348.0
\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.20348.0\\cppwinrt" /TcModules/_sha3/sha3module.c /Fobuild\temp.win-amd64-cpython-3
11\Release\Modules/_sha3/sha3module.obj
      sha3module.c
      C:\Users\baal\AppData\Local\Temp\pip-install-v45ogsce\pysha3_3b24b4b9e81c475baecd146d56e61954\Modules\_sha3\backport.inc(78): fatal error C10
83: 无法打开包括文件: “pystrhex.h”: No such file or directory
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\cl.exe' 
failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pysha3
  Running setup.py clean for pysha3
Failed to build pysha3
ERROR: Could not build wheels for pysha3, which is required to install pyproject.toml-based projects
(venv) PS F:\web3>

**Comment by dutterbutter** [comment]
hey @wt195799611 👋 

Thanks for posting this. I was able to reproduce on my side so we will definitely address this issue asap.

**Comment by danijelTxFusion** [comment]
I noticed that you are using Windows, so in order to make it work on Windows follow the [thread](https://github.com/zksync-sdk/zksync2-python/issues/3).

**Reply by dutterbutter** [reply]
It also fails on mac @danijelTxFusion :

```
Building wheels for collected packages: pysha3
  Building wheel for pysha3 (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-13-arm64-cpython-311
      copying sha3.py -> build/lib.macosx-13-arm64-cpython-311
      running build_ext
      building '_pysha3' extension
      creating build/temp.macosx-13-arm64-cpython-311
      creating build/temp.macosx-13-arm64-cpython-311/Modules
      creating build/temp.macosx-13-arm64-cpython-311/Modules/_sha3
      /opt/homebrew/opt/llvm/bin/clang -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk -I/opt/homebrew/opt/llvm/include -DPY_WITH_KECCAK=1 -I/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11 -c Modules/_sha3/sha3module.c -o build/temp.macosx-13-arm64-cpython-311/Modules/_sha3/sha3module.o
      In file included from Modules/_sha3/sha3module.c:20:
      Modules/_sha3/backport.inc:78:10: fatal error: 'pystrhex.h' file not found
      #include "pystrhex.h"
               ^~~~~~~~~~~~
      1 error generated.
      error: command '/opt/homebrew/opt/llvm/bin/clang' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pysha3
  Running setup.py clean for pysha3
Failed to build pysha3
ERROR: Could not build wheels for pysha3, which is required to install pyproject.toml-based projects
```

**Reply by danijelTxFusion** [reply]
On mac is the same problem, additional cpp library needs to be installed as it's explained in the thread.

**Comment by dutterbutter** [comment]
I was able to successfully install using python 3.10.

**Comment by henry412** [comment]
same problem，python version ：3.11

**Reply by dutterbutter** [reply, marked as answer]
You have to use python 3.10 for right now. General steps for macos:
```
Using Anaconda to setup a virtual environment with Python 3.10:
// Make the Python 3.10 project
mkdir ~/test/zksync-python-3-10-test
cd ~/test/zksync-python-3-10-test
conda create --name myenv python=3.10

// Activate the virtual environment
conda activate myenv

// Validate that you're running Python 3.10.12
python --version

// Install the zksync2 package
pip install zksync2
```

**Reply by baalisgood** [reply]
It never works when I use python3.10

**Reply by monkeyip** [reply]
Have you solved the problem? I still get this error using version 3.10

**Reply by monkeyip** [reply]
I have created the 3.10 virtual environment according to your method, but this error is still reported. How to di it?

**Comment by monkeyip** [comment]
> You have to use python 3.10 for right now. General steps for macos:
> 
> ```
> Using Anaconda to setup a virtual environment with Python 3.10:
> // Make the Python 3.10 project
> mkdir ~/test/zksync-python-3-10-test
> cd ~/test/zksync-python-3-10-test
> conda create --name myenv python=3.10
> 
> // Activate the virtual environment
> conda activate myenv
> 
> // Validate that you're running Python 3.10.12
> python --version
> 
> // Install the zksync2 package
> pip install zksync2
> ```

**Comment by biotremenag** [comment]
It seems like you're encountering an issue with installing the zksync2 package via pip due to the pysha3 dependency. The error message suggests that pysha3 has reached its end of life and is no longer supported, and it's prompting you to use SHA-3 functions from the hashlib module directly.

To resolve this issue, you can try the following steps:

Manually install the zksync2 package without including pysha3 as a dependency. You can do this by downloading the source code of zksync2, removing pysha3 from the dependencies list in its setup.py file, and then installing it using pip install ..

Alternatively, you can try installing an older version of pysha3 that is compatible with your Python version. You can search for a specific version of pysha3 that works with your environment and install it using pip install pysha3==<version>.

Consider reaching out to the maintainers of zksync2 to update their dependencies and remove the reliance on pysha3, especially since it has reached its end of life.

If you continue to encounter issues, you may need to explore alternative packages or solutions for interacting with zkSync using Python.

**Comment by davenportbri** [comment]
It seems that the error is occurring because the pysha3 package has reached its end of life and is no longer supported. Instead of using pysha3, you should use SHA-3 functions directly from the hashlib module in Python 3.6 and above. You may need to modify the zksync2 library to remove the dependency on pysha3 and use the appropriate SHA-3 functions from hashlib.

**Comment by loal0cksibus** [comment]
It seems that the pysha3 package is causing issues due to its end-of-life status and compatibility with Python 3.11. To resolve this, you can directly use SHA-3 functions from the hashlib module as recommended. This should allow you to proceed with your installation without relying on pysha3.
